### DISCUSSION TITLE
Displaying Posts and Threads

### DISCUSSION THREAD

**Discussion by CodedOre** [CodedOre/NewCaw#47] [Design]
As of now, I use the boxed list design with `Gtk.ListBox` for post timelines, and following the [original design](https://github.com/ibboard/cawbird/issues/177) it would be used for threads as well.

However, `Gtk.ListBox` has it's issues, especially with performance. The alternative would be to use `Gtk.ListView`, which automatically recycles it's used widgets on scrolling. The problem however is that it does not work with the boxed list design. The libadwaita theme currently discusses recommending a design based on the navigation sidebar.

Below is a concept for displaying a thread, left with the current design, right with a basic `ListView`-based one.

<img src="https://user-images.githubusercontent.com/47981497/182836485-5497f334-91b3-497b-853a-36c1b85f33d1.png" width="403px" height="800px"/><img src="https://user-images.githubusercontent.com/47981497/182837591-330d8ab8-1bba-4822-b045-085f9de26e6d.png" width="403px" height="800px"/>

<img src="https://user-images.githubusercontent.com/47981497/182902347-4effc2e1-a5cd-4e57-920a-c1d85f301c69.png" width="420px" height="311px"/><img src="https://user-images.githubusercontent.com/47981497/182902342-b785e5f8-bb28-4c7c-8270-19bf9ea06e96.png" width="420px" height="311px"/>

<img src="https://user-images.githubusercontent.com/47981497/182902315-c2afe94a-10c8-412b-b437-f56340dbe0d5.png" width="420px" height="311px"/><img src="https://user-images.githubusercontent.com/47981497/182902313-d79566e3-fa62-4920-b07d-29bb88b99424.png" width="420px" height="311px"/>

As I'm not sure yet about the best design, I want to open a discussion about how we ideally display these posts. The following points need to be considered:
- We have basic lists of unrelated posts in a timeline
  - Some timelines, for user and lists, are displayed under a detail view for said user or list, which currently uses the `card` style class, so would we need to change this too?
- We will also have a detail view of an post with the conversation around it
  - Unlike Cawbird 1.0, I want to display full threads, not only the immediate response, so we need to make clear which posts are part of an conversation. With the boxed list that was done with one card for each thread. With the list-style we could probably use lines between the posts like Twitter does.
  - We also should consider if we want to highlight the selected, "main" post in a coversation, and how.
- How do we display quotes?

**Comment by DylanVanAssche** [comment]
Not a designer, but I really like the right one, way more readable that it is a thread.

**Comment by IBBoard** [comment]
The right-hand one is cleaner. And performance is already a key complaint in Cawbird (although mainly when focussing/unfocussing the app, or when swapping between views).

I'm not sure how it would be in use when it's that bit darker, though. And I'm not sure that it separates the items quite enough. But maybe Adwaita has some options to improve that?

But I'd take it over the left-hand side having "cards" that are very white and obvious while the focussed/current tweet has less definition and is darker.

**Reply by CodedOre** [reply]
> I'm not sure how it would be in use when it's that bit darker, though.

Could you elaborate on that?

> And I'm not sure that it separates the items quite enough. But maybe Adwaita has some options to improve that?

Well, the thing is: `ListView`, rounded corners and separators are not really working together. So the libadwaita team attempts to circumvent separators. To quote [this comment](https://gitlab.gnome.org/GNOME/libadwaita/-/issues/516#note_1504978):

> - If we have separators, we can't have rounded rows as the .separators style uses border-radius.
> - Apps don't have direct control over the row widgets, which means we can't work around the above issue by using GtkSeparator or other methods.
>
> Meaning:
> 
> > Alternatively maybe divider lines between rows (or some other additional visual cue) could be a more general solution?
> 
> Does not work.

All in all, I am current porting to the new style on the [`list-post`](https://github.com/CodedOre/NewCaw/tree/list-post) branch.

**Reply by IBBoard** [reply]
>>    I'm not sure how it would be in use when it's that bit darker, though.
>
> Could you elaborate on that?

Sorry, that was a poorly constructed sentence where I missed some words through lazy use of English.

I'm not sure how the slightly darker grey would feel in day-to-day use. My initial impression is that the brighter background on the left has more contrast and will be easier to read for longer periods of time (to the extent that Adwaita is usable for anyone - personally I feel it's too bright _but_ the white is the normal "text boxes for you to read" background and my custom theme darkens it a bit, which could then become too dark in the right-hand theme)

> Well, the thing is: ListView, rounded corners and separators are not really working together. So the libadwaita team attempts to circumvent separators. To quote [this comment](https://gitlab.gnome.org/GNOME/libadwaita/-/issues/516#note_1504978):
> …

Also from that discussion: "The goal here is to stop apps from trying to do boxed list views. It never works and we don't want more apps to do what Music did there." - which is a shame, because the people saying that it loses clarity and grouping when widgets aren't bounded in some way _are right_.

**Reply by CodedOre** [reply]
> I'm not sure how the slightly darker grey would feel in day-to-day use. My initial impression is that the brighter background on the left has more contrast and will be easier to read for longer periods of time (to the extent that Adwaita is usable for anyone - personally I feel it's too bright but the white is the normal "text boxes for you to read" background and my custom theme darkens it a bit, which could then become too dark in the right-hand theme)

That is a valid point! We could use the `view`-style as an background. My main problem with it (as of now) is that then there is not a lot of contrast to the `card`-styled user information box, hence the one point in the initial comment. I will look into that tomorrow.

> Also from that discussion: "The goal here is to stop apps from trying to do boxed list views. It never works and we don't want more apps to do what Music did there." - which is a shame, because the people saying that it loses clarity and grouping when widgets aren't bounded in some way are right.

I think this is only meant for the times when an app uses `Gtk.ListView`, because there boxed lists don't really work.
